### ISSUE TITLE
DOC: ufunc instructions use deprecated distutils

### ISSUE THREAD

**Issue by oscarbenjamin** [numpy/numpy#22027]
### Issue with current documentation:

While working on reducing warnings from SymPy I found myself looking at this page of the NumPy docs:
https://numpy.org/doc/stable/user/c-info.ufunc-tutorial.html

The code examples from there seem to be used more or less directly in SymPy:
https://github.com/sympy/sympy/blob/39de9a2698ad4bb90681c0fdb70b30a78233145f/sympy/utilities/autowrap.py#L702-L801

I found warnings from running this because of the usage of distutils and numpy.distutils. Since both types of distutils are deprecated I guess that the appropriate update would be to use setuptools.

Also compiling the examples from the docs page gives a warning:
```
/usr/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it by " \
  ^
```

### Idea or request for content:

For me on Ubuntu it seems to work fine to just use this `setup.py` with setuptools:
```python
from setuptools import setup, Extension

npufunc = Extension('npufunc', sources=['single_type_logit.c'])

setup(name = 'spam',
    version='1.0',
    description='This is my spam package',
    ext_modules = [npufunc])
```
That would seem the most straight-forward update to the docs.

I don't know though if setuptools can be expected to work like this on all platforms though and I'm also not really sure why the docs pages suggests using `numpy.distutils` at all so it's not clear to me that setuptools can replace it here.

Another possibility would be to add instructions showing how to do the same using other build tools rather than setuptools.

The other warning can be addressed by adding the line
```c
#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
```

**Comment by rgommers** [comment]
Thanks for the report @oscarbenjamin 

> I found warnings from running this because of the usage of distutils and numpy.distutils. Since both types of distutils are deprecated I guess that the appropriate update would be to use setuptools.

It's not, we plan to follow SciPy in moving to Meson. Initial work has started, but that'll take a while. For now, just ignore the warnings. `distutils` and `numpy.distutils` are indeed deprecated, the warnings are there for a reason. The main point is not to use them directly if you can avoid it.

> Also compiling the examples from the docs page gives a warning:

That indeed can and should be fixed now, it's not hard to do. Your suggestion to add `NPY_NO_DEPRECATED_API` seems right.

> I don't know though if setuptools can be expected to work like this on all platforms though and I'm also not really sure why the docs pages suggests using `numpy.distutils` at all so it's not clear to me that setuptools can replace it here.

Probably because `numpy.distutils` is older than `setuptools`, and it's what numpy itself uses. 

> For me on Ubuntu it seems to work fine to just use this `setup.py` with setuptools:

That would be fine to do now.

**Comment by oscarbenjamin** [comment]
> > I guess that the appropriate update would be to use setuptools.
> 
> It's not, we plan to follow SciPy in moving to Meson

I guess you're talking about the build system for building NumPy itself. I'm talking about what is suggested in the docs here for users. To me it seems that these simple examples can be built with setuptools which is potentially simpler than meson (I don't know that though because I haven't used meson).

In SymPy's case this is not part of building SymPy but rather a feature that allows users to create a ufunc from a symbolic expression. SymPy will generate a C extension module and compile it under the hood:
```python
In [2]: from sympy.utilities.autowrap import ufuncify

In [3]: x = symbols('x')

In [4]: logit = sympy.log((1 - x)/x)

In [5]: logit
Out[5]: 
   ⎛1 - x⎞
log⎜─────⎟
   ⎝  x  ⎠

In [7]: f = ufuncify(x, logit)

In [8]: f([0.1, 0.2, 0.3])
Out[8]: array([2.19722458, 1.38629436, 0.84729786])
```
Here the build system needs to exist in the user's environment and they are the one who would see the warning (or error in future) as a result of the usage of `numpy.distutils`. It seems to me that the safest replacement for `distutils` and `numpy.distutils` here is `setuptools` rather than `meson` which I imagine is unlikely to be in the user's environment at runtime. In any case this is an optional feature so SymPy will not have a hard dependency on `setuptools`, `meson` or anything else.

**Comment by eli-schwartz** [comment]
> I don't know though if setuptools can be expected to work like this on all platforms though and I'm also not really sure why the docs pages suggests using `numpy.distutils` at all so it's not clear to me that setuptools can replace it here.

Your setup.py using pure setuptools fails for me with the following *highly* informative error message:

```
running build
running build_ext
error: Command "gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -flto -ffat-lto-objects -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -flto -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -flto -fPIC -I/usr/include/python3.10 -c single_type_logit.c -o build/temp.linux-x86_64-3.10/single_type_logit.o" failed with exit status 1
```

"failed with exit status 1", oh joy. :joy: Manually running that command shows the error message that setuptools deleted from the output: `single_type_logit.c:3:10: fatal error: numpy/ndarraytypes.h: No such file or directory`

This happens because numpy.distutils adds the numpy include directories when compiling the extension, but setuptools does not. It nevertheless works on Ubuntu, because Ubuntu symlinks the numpy include directory so that it's on the default include path, but I'd say this is really worse than doing nothing, because if it doesn't work *anywhere* then that's obvious and people fix it, but this Ubuntu-ism will cause "it work on my machine" situations and then projects have to field bug reports from people on non-Debianlike systems who get very confused...

Anyway, adding `include_dirs=[np.get_include()]` suffices there. A similar incantation is used in e.g. SciPy's meson.build (though there's a ticket to make this a dependency class).

You don't really need numpy.distutils I guess... it just happens to be convenient in this case even when you're not using any of the advanced features setuptools doesn't support.

> It seems to me that the safest replacement for `distutils` and `numpy.distutils` here is `setuptools` rather than `meson` which I imagine is unlikely to be in the user's environment at runtime. In any case this is an optional feature so SymPy will not have a hard dependency on `setuptools`, `meson` or anything else.

As a build-time replacement, no one needs to worry about what's likely to be in the user environment, since either the user will install it per the build-time documentation, or pip will install it automatically.

As a runtime replacement, I guess you might want to implement the most minimal build system you can, yourself, and directly call the C compiler.

**Comment by rgommers** [comment]
> Here the build system needs to exist in the user's environment and they are the one who would see the warning (or error in future) as a result of the usage of `numpy.distutils`. It seems to me that the safest replacement for `distutils` and `numpy.distutils` here is `setuptools` rather than `meson` which I imagine is unlikely to be in the user's environment at runtime. In any case this is an optional feature so SymPy will not have a hard dependency on `setuptools`, `meson` or anything else.

Makes sense. It can be replaced here with `setuptools` for now. The best pattern we have for this is to simply give info for multiple build systems, like in https://numpy.org/devdocs/f2py/buildtools/index.html#build-systems - so bonus points if someone wants to do that here (it's not too hard, since it's only a few lines of code).

**Comment by oscarbenjamin** [comment]
> Anyway, adding `include_dirs=[np.get_include()]` suffices there

Thanks, that's precisely the sort of thing I was worried about. I knew that `np.get_include()` used to be needed in some cases but was surprised that it seemed to work without. The first example in the docs page even has `include_dirs=['/usr/local/lib']` which doesn't look portable and doesn't even seem to be necessary on my system where that directory does exist.

Hopefully then the correct setuptools replacement for the setup.py shown on that docs page is:
```python
from setuptools import setup, Extension
import numpy as np

npufunc = Extension('npufunc',
                    sources=['single_type_logit.c'],
                    include_dirs=[np.get_include()])

setup(name = 'spam',
    version='1.0',
    description='This is my spam package',
    ext_modules = [npufunc])
```

**Comment by rgommers** [comment]
That looks right

**Comment by rgommers** [comment]
Completed in gh-26566, so closing. Thanks all!
